File: Python/redundant_connection_ii.py
def findRedundantDirectedConnection(edges: list) -> list:
    def find(index: int) -> int:
        if ancestor[index] != index:
            ancestor[index] = find(ancestor[index])
        return ancestor[index]

    def union(index1: int, index2: int):
        ancestor[find(index1)] = find(index2)

    nodes_count = len(edges)
    ancestor = list(range(nodes_count + 1))
    parent = list(range(nodes_count + 1))
    conflict = -1
    cycle = -1
    for i, (node1, node2) in enumerate(edges):
        if parent[node2] != node2:
            conflict = i
        else:
            parent[node2] = node1
            if find(node1) == find(node2):
                cycle = i
            else:
                union(node1, node2)
    if conflict < 0:
        return [edges[cycle][0], edges[cycle][1]]
    else:
        conflict_edge = edges[conflict]
        if cycle >= 0:
            return [parent[conflict_edge[1]], conflict_edge[1]]
        else:
            return [conflict_edge[0], conflict_edge[1]]

if __name__ == "__main__":
    print(findRedundantDirectedConnection([[1,2], [2,3], [3,4], [4,1], [1,5]]))

File: src/tmp/Java/src/com/dbc/algorithm/UnionFindSets/RedundantConnectionII.java
package com.dbc.algorithm.UnionFindSets;

public class RedundantConnectionII {
    public static int[] findRedundantDirectedConnection(int[][] edges) {
        int[] ancestor = new int[edges.length + 1];
        int[] parent = new int[edges.length + 1];

        for (int i = 0; i <= edges.length; i++) {
            ancestor[i] = i;
            parent[i] = i;
        }

        int conflict = -1;
        int cycle = -1;

        for (int i = 0; i < edges.length; i++) {
            int node1 = edges[i][0];
            int node2 = edges[i][1];
            if (parent[node2] != node2) {
                conflict = i;
            } else {
                parent[node2] = node1;
                if (find(ancestor, node1) == find(ancestor, node2)) {
                    cycle = i;
                } else {
                    union(ancestor, node1, node2);
                }
            }
        }

        if (conflict < 0) {
            return edges[cycle];
        } else {
            if (cycle >= 0)
                edges[conflict][0] = parent[edges[conflict][1]];
            return edges[conflict];
        }
    }

    public static int find(int[] ancestor, int index) {
        if (ancestor[index] != index) {
            ancestor[index] = find(ancestor, ancestor[index]);
        }
        return ancestor[index];
    }

    public static void union(int[] ancestor, int index1, int index2) {
        ancestor[find(ancestor, index1)] = find(ancestor, index2);
    }
}
